cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -305,76 +327,173 @@ func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
+func (h eventsRPCHandler) getEvents(ctx context.Context, request GetEventsRequest) (GetEventsResponse, error) {
 	if err := request.Valid(h.maxLimit); err != nil {
 		return GetEventsResponse{}, &jrpc2.Error{
-			Code:    jrpc2.InvalidParams,
-			Message: err.Error(),
+			Code: jrpc2.InvalidParams, Message: err.Error(),
 		}
 	}
 
-	start := events.Cursor{Ledger: uint32(request.StartLedger)}
+	ledgerRange, err := h.ledgerReader.GetLedgerRange(ctx)
+	if err != nil {
+		return GetEventsResponse{}, &jrpc2.Error{
+			Code: jrpc2.InternalError, Message: err.Error(),
+		}
+	}
+
+	start := db.Cursor{Ledger: request.StartLedger}
 	limit := h.defaultLimit
 	if request.Pagination != nil {
 		if request.Pagination.Cursor != nil {
 			start = *request.Pagination.Cursor
-			// increment event index because, when paginating,
-			// we start with the item right after the cursor
+			// increment event index because, when paginating, we start with the item right after the cursor
 			start.Event++
 		}
 		if request.Pagination.Limit > 0 {
 			limit = request.Pagination.Limit
 		}
 	}
+	endLedger := request.StartLedger + LedgerScanLimit
 
-	type entry struct {
-		cursor               events.Cursor
-		ledgerCloseTimestamp int64
-		event                xdr.DiagnosticEvent
-		txHash               *xdr.Hash
-	}
-	var found []entry
-	latestLedger, err := h.scanner.Scan(
-		events.Range{
-			Start:      start,
-			ClampStart: false,
-			End:        events.MaxCursor,
-			ClampEnd:   true,
-		},
-		func(event xdr.DiagnosticEvent, cursor events.Cursor, ledgerCloseTimestamp int64, txHash *xdr.Hash) bool {
-			if request.Matches(event) {
-				found = append(found, entry{cursor, ledgerCloseTimestamp, event, txHash})
-			}
-			return uint(len(found)) < limit
-		},
-	)
+	if request.EndLedger != 0 {
+		endLedger = min(request.EndLedger, endLedger)
+	}
+
+	end := db.Cursor{Ledger: endLedger}
+	cursorRange := db.CursorRange{Start: start, End: end}
+
+	if start.Ledger < ledgerRange.FirstLedger.Sequence || start.Ledger > ledgerRange.LastLedger.Sequence {
+		return GetEventsResponse{}, &jrpc2.Error{
+			Code: jrpc2.InvalidRequest,
+			Message: fmt.Sprintf(
+				"startLedger must be within the ledger range: %d - %d",
+				ledgerRange.FirstLedger.Sequence,
+				ledgerRange.LastLedger.Sequence,
+			),
+		}
+	}
+
+	found := make([]entry, 0, limit)
+
+	contractIDs, err := combineContractIDs(request.Filters)
+	if err != nil {
+		return GetEventsResponse{}, &jrpc2.Error{
+			Code: jrpc2.InvalidParams, Message: err.Error(),
+		}
+	}
+
+	topics, err := combineTopics(request.Filters)
+	if err != nil {
+		return GetEventsResponse{}, &jrpc2.Error{
+			Code: jrpc2.InvalidParams, Message: err.Error(),
+		}
+	}
+
+	eventTypes := combineEventTypes(request.Filters)
+
+	// Scan function to apply filters
+	eventScanFunction := func(
+		event xdr.DiagnosticEvent, cursor db.Cursor, ledgerCloseTimestamp int64, txHash *xdr.Hash,
```

Comment:
why is txHash by reference? maybe event should be too? might save memory (less ram) but decrease locality (more cpu) idk

Reply:
Yeah we could even do that with `event` obj to save some memory but allocs and memory usage was already low so was thinking to do it in follow up PR after seeing some traffic and stats around that in prod.